src/PasswordHistory.php

```diff
@@ -0,0 +1,207 @@
+        }
+
+        // Check if password history is enabled
+        if ($length == 0) {
+            // No password history, no need to go further
+            return true;
+        }
+
+        $passwords_checked_count = 0;
+        foreach ($this->getForUser($user) as $previous_password) {
+            if (Auth::checkPassword($password, $previous_password)) {
+                // New password match an entry in the history
+                return false;
+            }
+
+            // History might contain more entries than configurated (skip them)
+            $passwords_checked_count++;
+            if ($passwords_checked_count == $length) {
+                break;
+            }
+        }
+
+        // New password didn't match any entry in the history
+        return true;
+    }
+
+    /**
+     * Update password history for a given user
+     * Must be called after a user password is updated
+     *
+     * @param User $user
+     *
+     * @return void
+     */
+    public function updatePasswordHistory(User $user): void
+    {
+        $password = $user->oldvalues['password'] ?? '';
+
+        // No password change
+        if (empty($password)) {
+            return;
+        }
+
+        // Here $password should always be a hash and not a "real" password
+        // To be extra safe and make sure we don't log any real password we can
+        // verify that the hash algorithm is known
+        if (password_get_info($password)['algoName'] == 'unknown') {
+            trigger_error(
+                "Unhashed password passed to PasswordHistory->updatePasswordHistory()",
+                E_USER_WARNING
+            );
+            return;
+        }
+
+        // Get full history
+        $history = $this->getForUser($user, true);
+
+        // Add hash to history
+        array_unshift($history, $password);
+
+        // Clear extra values in history
+        $history = array_slice($history, 0, self::MAX_HISTORY_SIZE);
+
+        // Save to DB
+        $update = $user->update([
+            'id' => $user->getId(),
+            'password_history' => exportArrayToDB($history),
+        ]);
+
+        // Should not happen - log warning
+        if (!$update) {
+            trigger_error(
+                "Password history update failed for user {$user->getId()}",
+                E_USER_WARNING
+            );
+        }
+    }
+
+    /**
+     * Get the password history of a given user
+     *
+     * @param User $user
+     * @param bool $full
+     *
+     * @return array
+     */
+    public function getForUser(User $user, bool $full = false): array
+    {
+        $values = importArrayFromDB($user->fields['password_history']);
+
+        if (!$full) {
+            $limit = min($this->getConfiguratedLength(), self::MAX_HISTORY_SIZE);
+            $values = array_slice($values, 0, $limit);
+        }
+
+        return $values;
+    }
+
+    /**
+     * Singleton constructor
+     */
+    private function __construct()
+    {
+    }
```

Comment:
This is useless.
```suggestion
```

Reply:
Singleton constructors are supposed to be private.
Its not that important but can't hurt applying the pattern correctly (even if its a shit pattern that shouldn't really be used for this in a perfect world).

Reply:
I did not see it was private. My bad.